Add the UpgradeNotice to the Calendly and OpenTable blocks

scruffian opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Fixes #14743

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* When viewing the block on a site in which the block it is not available, display and upgrade notice with a button

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Bug fix

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Apply the phab patch to your sandbox
* Add a Calendly block and and OpenTable block to a free Simple site
* Open the front end of the site as an admin
* Check that you see a notice for both blocks like this:
<img width="728" alt="Screenshot 2020-03-11 at 16 34 01" src="https://user-images.githubusercontent.com/275961/76440690-2427b080-63b6-11ea-93c6-85101b318e84.png">

Note, your preview may not look like this, if you have an older version of the components package installed. See https://github.com/Automattic/jetpack/pull/14934

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* no changelog

scruffian committed:
Add the UpgradeNotice to the Calendly and OpenTable blocks

Copons reviewed (dismissed):
LGTM!

I think we might follow up (or update this) by moving the availability functions into a shared utility to avoid code duplication.
(AFAICS they are pretty much identical in Calendly and OpenTable)

jeherve commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -63,7 +62,7 @@ function set_availability() {
 			array(
 				'required_feature' => 'opentable',
-				'required_plan'    => 'premium-plan',
+				'required_plan'    => 'value_bundle',
```
Should it use `REQUIRED_PLAN` here?

```suggestion
				'required_plan'    => REQUIRED_PLAN,
```

jeherve commented on extensions/blocks/calendly/calendly.php:
```diff
@@ -63,7 +62,7 @@ function set_availability() {
 			'missing_plan',
 			array(
 				'required_feature' => 'calendly',
```
Maybe we can take that opportunity to update this too?

```suggestion
				'required_feature' => FEATURE_NAME,
```

jeherve commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -63,7 +62,7 @@ function set_availability() {
 			'missing_plan',
 			array(
 				'required_feature' => 'opentable',
```
```suggestion
				'required_feature' => FEATURE_NAME,
```

jeherve commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -41,12 +42,10 @@ function is_available() {
-	}
+	jetpack_register_block(
+		BLOCK_NAME,
+		array( 'render_callback' => 'Jetpack\OpenTable_Block\load_assets' )
```
Let's take the opportunity to update this:

```suggestion
		array( 'render_callback' => __NAMESPACE__ . '\load_assets' )
```

jeherve commented on extensions/blocks/calendly/calendly.php:
```diff
@@ -41,12 +42,10 @@ function is_available() {
-	}
+	jetpack_register_block(
+		BLOCK_NAME,
+		array( 'render_callback' => 'Jetpack\Calendly_Block\load_assets' )
```
```suggestion
		array( 'render_callback' => __NAMESPACE__ . '\load_assets' )
```

jeherve commented on extensions/blocks/calendly/calendly.php:
```diff
@@ -79,9 +78,14 @@ function set_availability() {
  * @return string
  */
 function load_assets( $attr, $content ) {
+	if ( ! is_available() ) {
```
For Simple Payments, we checked if we were not on an RSS feed: 
https://github.com/Automattic/jetpack/blob/7f3bd724c184747db76937e367d1840c7a91fb6f/modules/simple-payments/simple-payments.php#L159

See #13723

Maybe we should do the same here, or integrate that check to the `upgrade_nudge` method?

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -937,4 +937,26 @@ public static function validate_block_embed_url( $url, $allowed = array(), $is_r
+	 *
+	 * @return string
+	 */
+	public static function upgrade_nudge( $plan ) {
```
I wonder if we should add the 2 filters we use here:
https://github.com/Automattic/jetpack/blob/557a85a6a2b0c9b5dc1535212632904255fda1e3/class.jetpack-gutenberg.php#L254-L267

Those filters are there to hide the nudges in Jetpack, I think we should use them for frontend nudges as well.

jeherve reviewed (changes requested):
This seems like a good opportunity to update Simple Payments maybe?
https://github.com/Automattic/jetpack/blob/7f3bd724c184747db76937e367d1840c7a91fb6f/modules/simple-payments/simple-payments.php#L179-L189

apeatling committed:
Add fixes to upgrade nudge function and block based constants needed after PR code review.

apeatling commented:
Added changes needed based on @jeherve's feedback above. Needs a re-review.

scruffian commented on class.jetpack-gutenberg.php:
```diff
@@ -947,7 +947,10 @@ public static function validate_block_embed_url( $url, $allowed = array(), $is_r
+		/** This filter is documented in _inc/lib/admin-pages/class.jetpack-react-page.php */
+		if ( ! apply_filters( 'jetpack_block_editor_enable_upgrade_nudge', false )
+			|| ! apply_filters( 'jetpack_show_promotions', true )
+			|| is_feed() ) {
```
Do we need to leave `manage_options` too, so that this only shows for the admin of the site?

apeatling replied:
Yes! My mistake, will add this back in.

apeatling replied:
✅

apeatling committed:
Restore missing manage_options check.

Copons commented:
> Note, your preview may not look like this, if you have an older version of the components package installed. See #14934

Just mentioning that indeed the upgrade notice would look like this (second Upgrade button is hovered):
<img width="620" alt="Screenshot 2020-03-17 at 16 35 34" src="https://user-images.githubusercontent.com/2070010/76878920-6a28bc80-686d-11ea-980c-f87742afc9c3.png">

Copons commented on class.jetpack-gutenberg.php:
```diff
@@ -937,4 +937,30 @@ public static function validate_block_embed_url( $url, $allowed = array(), $is_r
+	 * @return string
+	 */
+	public static function upgrade_nudge( $plan ) {
+		/** This filter is documented in _inc/lib/admin-pages/class.jetpack-react-page.php */
```
I find this comment a bit confusing.
The following line is not a filter, and the first filter appearing (`jetpack_block_editor_enable_upgrade_nudge`), although used in that file there, is in fact documented here (line 251)! 😄 

The comment seems to refer to `jetpack_show_promotions`, so I'd probably either move the comment down, or clarifiy it.

jeherve replied:
Yeah, this isn't going to work I am afraid, the parser won't pick it up. We'd need to move the comment right above its filter. We'll need a similar one for `jetpack_block_editor_enable_upgrade_nudge`:

`/** This filter is documented in class.jetpack-gutenberg.php */`

(even if we're in the same file. That will help the parser understand that no page should be created for that hook).

Something like this:

```diff
diff --git a/class.jetpack-gutenberg.php b/class.jetpack-gutenberg.php
index 577fa4e8b..d7759ec8c 100644
--- a/class.jetpack-gutenberg.php
+++ b/class.jetpack-gutenberg.php
@@ -947,11 +947,14 @@ class Jetpack_Gutenberg {
 	 * @return string
 	 */
 	public static function upgrade_nudge( $plan ) {
-		/** This filter is documented in _inc/lib/admin-pages/class.jetpack-react-page.php */
-		if ( ! current_user_can( 'manage_options' )
+		if (
+			! current_user_can( 'manage_options' )
+			/** This filter is documented in class.jetpack-gutenberg.php */
 			|| ! apply_filters( 'jetpack_block_editor_enable_upgrade_nudge', false )
+			/** This filter is documented in _inc/lib/admin-pages/class.jetpack-react-page.php */
 			|| ! apply_filters( 'jetpack_show_promotions', true )
-			|| is_feed() ) {
+			|| is_feed()
+		) {
 			return;
 		}
 

```

Copons reviewed (dismissed):
Confusing comment aside, this works as advertised.

I also took care of rebasing D40191-code that was stuck at the first commit.

jeherve reviewed (changes requested):
I only have a minor comment to avoid issues with the Codex parser. 

I still think we could take that opportunity to update the Simple Payments block as well:
https://github.com/Automattic/jetpack/pull/14953#pullrequestreview-373418025

If not in this PR, do you think you could create an issue so this doesn't fall off our radar?

Thank you!

scruffian committed:
update the comment

scruffian commented:
I updated the comment and rebased the patch. This is ready for another test and hopefully a merge! I'll do the simple payments block in a different PR just because this one is so stale...

I created an issue for that here: https://github.com/Automattic/jetpack/issues/15591

jeherve reviewed (approved):
This tests well for me. 👍

kraftbj commented:
r206747-wpcom
